### PR TITLE
Return more correct code/debug identifiers for macOS minidumps

### PR DESCRIPTION
**Pull request by relaxolotl**
This is related to the initiative outlined in https://github.com/getsentry/symbolicator/issues/620.

This PR fills in the currently incomplete implementation of `code_identifier` as well as a subsection of `debug_identifier` on macOS/Solaris(/maybe iOS, NaCl, PS3) minidump modules. The way this is done is by relying on the assumption that the `age` field on a module's CV record is only ever `0` if the minidump was created by Breakpad on a non-Windows device. See Breakpad's minidump writer for [macOS](https://github.com/google/breakpad/blob/main/src/client/mac/handler/minidump_generator.cc#L1468-L1486) and [Solaris](https://github.com/google/breakpad/blob/main/src/client/solaris/handler/minidump_generator.cc#L474-L488).

The reason why I'm depending on this rather fragile behaviour is because the correct way of going about this would [require access to the contents of a `MinidumpSystemInfo` in `code_identifier` and possibly `debug_identifier`](https://github.com/google/breakpad/blob/main/src/processor/minidump.cc#L1988-L1999). As far as I can tell, doing this the correct way would require much more invasive changes to the API, so I went with this hacky approach to get the ball rolling.

Some of the aforementioned invasive changes involve changing the `Module` trait in `minidump-common`, or adding an `Option<MinidumpSystemInfo>` field to a `MinidumpModule`. I've taken a look at `get_crash_reason` but it doesn't completely neatly map to the use case we have here. The way `MinidumpThread` works with stream dependencies looked like a good start, though - would appreciate thoughts on this.

Given that it's likely that a sizable amount of the diff here will be affected by https://github.com/luser/rust-minidump/pull/432, what I'd really love feedback on here is whether the approach here is going in the right direction or if I should re-think how to solve this problem.

### CODE REVIEW

**Comment by flub** [comment]
It's kind of hard to reason about on its own.  The symbolicator test suite has a macos minidump I wonder if we can extract some debugIDs from there to add to some tests here?

**Comment by flub** [comment]
> It's kind of hard to reason about on its own. The symbolicator test suite has a macos minidump I wonder if we can extract some debugIDs from there to add to some tests here?

oh wait, i should actually read the full PR 😓

File: minidump/src/minidump.rs
Comment (Gankra) [inline minidump/src/minidump.rs:923]: Any chance you could make these links canonicalized?

File: minidump/src/minidump.rs
Comment (Gankra) [inline minidump/src/minidump.rs:929]: We have a bunch of endianess machinery between `scroll` and the stdlib, can we not leverage that here to do this a bit less messily..?

**Review by Gankra — CHANGES_REQUESTED** [review]
Some nits

**Comment by Gankra** [comment]
Note also that I have established a precedent for certain getters requiring extra state to continue lazily parsing the value: https://github.com/luser/rust-minidump/blob/8e1f9a8f9fb67d25a817b48fdee7b9870e1cb321/minidump/src/minidump.rs#L2217-L2221

I can certainly respect not wanting to introduce that here, but it's worth considering longterm.

**Comment by Gankra** [comment]
Ah I see you've seen a bunch of this -- I should *also* read the full PR lol. I'm gonna take a closer look at https://github.com/luser/rust-minidump/pull/432 first

File: minidump/src/minidump.rs
Comment (relaxolotl) [reply minidump/src/minidump.rs:929]: Were there specific helpers that you had in mind that could be used to do this a little more elegantly? 

There's also a bit of ambiguity here that I'd like to clarify: Are you asking that the
1. Entire GUID is re-read with the correct endianness
2. `data1-3` are swapped using `scroll` instead of `swap_bytes`?

Comment (relaxolotl) [reply minidump/src/minidump.rs:929]: fwiw it looks like https://github.com/luser/rust-minidump/pull/438 also adds the uuid crate as a dependency, which opens up the option of using [`Uuid::from_fields_le`](https://docs.rs/uuid/0.8.2/uuid/struct.Uuid.html#method.from_fields_le) that appears to be designed for the exact conversion that's being performed in this block. Maybe that's the best option to go with here?

**Comment by gabrielesvelto** [comment]
> Also: I would not give any consideration to platforms where you don't have extant minidumps to work with (like Solaris, NaCl, PS3). They're mostly dead anyway, and if someone pops up wanting support for those platforms the burden will be on that person to demonstrate what the right thing to do is.

:point_up: this. We explicitly avoided supporting older architectures in all the breakpad code we replaced. It just adds to the maintenance burden and has no real upside.

**Comment by Swatinem** [comment]
Well, I’m absolutely do not look forward to having to maintain that stuff, but we even had to fix up the breakpad mips client code recently on behalf of customers 😅

**Comment by relaxolotl** [comment]
I hooked this change up to symbolicator just to sanity check these changes and it looks like the GUID byte flipping actually isn't needed - it actually mucks up the UUIDs on macOS dumps. It doesn't look like Breakpad did all of swapping to begin with when reading `debug_identifier`s for macOS, so I guess this lines up. Truth be told, I'm a little confused about these GUIDs/UUIDs and when and where these need to be manually flipped, so I'll be trusting the tests here and on symbolicator.

> I would not give any consideration to platforms where you don't have extant minidumps to work with (like Solaris, NaCl, PS3).

I've stripped away references to those platforms, thanks for the heads up! 

> Given that the Mach-O case is not, I feel like maybe there's a more sensible path forward here?

Could I bug you for a bit more detail on this? I'm afraid I'm not too sure what this count be hinting at, probably due to my lack of familiarity with things - Was there something specific you had in mind, were you thinking that changes to breakpad upstream would be better, or?

File: minidump/src/minidump.rs
Comment (Gankra) [reply minidump/src/minidump.rs]: That sounds good, let's wait until that patch lands

Comment (relaxolotl) [reply minidump/src/minidump.rs]: So I did some testing with symbolicator and it looks like we didn't need this swapping after all. This gets rid of the ugly bits pointed out in the original comment and removes the loose dependency this PR has on #438. 🎉

**Comment by relaxolotl** [comment]
Polished this up a bit more, but I think this is otherwise good for a final review. I spoke to @flub a bit about this and we'll sort out any rebasing/merge conflicts that's needed on #432 instead of delaying this PR since #432's looking like a more involved change. This should remove any dependency this PR has to #432.

I'm assuming that it's okay to hold off on figuring out how to make `SystemInfo` accessible to `debug_identifier` and `code_identifier` in this PR (maybe I can open up an issue to track this?)

@luser you'd mentioned in your comment that you'd take a look over this - based on the activity in the PR I'm assuming that you haven't had a chance to do so yet. Is that correct, or could I be getting things completely wrong here?

cc @Gankra since you reviewed this 🙏

File: minidump/src/minidump.rs
Comment (mitsuhiko) [reply minidump/src/minidump.rs]: The microsoft PDB implementation has this to say: https://github.com/microsoft/microsoft-pdb/blob/082c5290e5aff028ae84e43affa8be717aa7af73/PDB/dbi/pdb.cpp#L602-L606

> we reserve 0 as a special age value and start counting from 1

**Review by flub — APPROVED** [review]

**Comment by Gankra** [comment]
Ok, now that the big debugid PR landed, I'll look at this more seriously.

**Review by luser — APPROVED** [review]
OK, this seems fine given the due diligence about `age` starting at 1 for MSVC-produced PDB files.

My offhand comment that you asked for clarification on:
> Given that the Mach-O case is not, I feel like maybe there's a more sensible path forward here?

My thinking there was that if you know this is a minidump written on macOS/iOS, you know exactly what the format of the data is. The "more sensible path" here would be to do the same system info OS checks that Breakpad does. I'm not terribly concerned about that right now given that code identifiers are not a load-bearing part of crash report processing, and this change makes it so that this code will produce sensible values for all of Windows, macOS/iOS, and Linux/Android.

**Comment by luser** [comment]
A more-thorough approach to this would perhaps implement a Mach-O equivalent to the [`CV_INFO_ELF`](https://docs.rs/minidump/latest/minidump/format/struct.CV_INFO_ELF.html) I implemented. One could imagine adding something like:

```rust
pub struct CV_INFO_MACHO_UUID {
  pub cv_signature: u32, // Make up a CvSignature::MachO
  pub uuid: GUID,            // directly from the LC_UUID
}
```

Defining such a thing and implementing support for it in this codebase would be straightforward. However, that's not very useful without changing minidump writing clients to use it:

- Changing [Firefox's fork of the mac Breakpad client](https://searchfox.org/mozilla-central/rev/3de1b6791ea569cdf7773f7cd512cf2e6cc1d3f0/toolkit/crashreporter/breakpad-client/mac/handler/minidump_generator.cc#1480-1583) would not be too bad.
- Changing [the same code in the mac Breakpad client](https://chromium.googlesource.com/breakpad/breakpad/+/f6974b15ef96614e9a7df25c3ff31e350bd4e346/src/client/mac/handler/minidump_generator.cc#1424) would presumably require adding support for reading this data to the Breakpad processor.
- I looked at the crashpad source for this and I have regrets, but I suppose it could be changed. It would likely be more productive to convince someone at Google to do the work:
  - [The mac module snapshot code that fetches the UUID](https://source.chromium.org/chromium/chromium/src/+/main:third_party/crashpad/crashpad/snapshot/mac/module_snapshot_mac.cc;l=161;drc=b49a213e324277a5539cc808e46cfa982502a81d)
  - [The module writing code that determines what type of CV record to write](https://source.chromium.org/chromium/chromium/src/+/main:third_party/crashpad/crashpad/minidump/minidump_module_writer.cc;l=279;drc=c26af9bfc075bbed0c70d2371b9021d3a3479d92) is…not structured in a way that I would have chosen if I had written it.

For historical context:

- [This Breakpad commit](https://chromium.googlesource.com/breakpad/breakpad/+/4912669df1fb34fb8549925a8f7e70e5ffc9e890) is where I added the `CV_INFO_ELF` structure and support for reading it in the processor code.
- [This later commit](https://chromium.googlesource.com/breakpad/breakpad/+/6c8f80aa8b3ba8120c4158c069bb298c044dedf9) is where I changed the Linux minidump writer to write CV records in that format.

File: minidump/src/minidump.rs
Comment (relaxolotl) [reply minidump/src/minidump.rs]: Good catch, they are indeed the same. I've also added a comment noting why there isn't a special case for macOS here just to be safe.

**Comment by relaxolotl** [comment]
Thanks for the in-depth explanation, @luser; I see what you mean by the "more sensible" path. I could take a crack at solving it that way (figuring out a way to read the system info) in a follow-up PR, if you feel that this is best figured out sooner than later. It's not clear to me what an API for that would look like though, so chances are I'd need to have a bit of back-and-forth regarding what such a solution should look like before getting started on the code changes.

The more thorough approach sounds like a pretty massive undertaking. It's pretty unlikely that I'll be tackling this (heck, I'm not sure where I'd get started talking to somebody at Google about changing crashpad) but I'd hate to see this buried in this PR. Thoughts?

**Comment by relaxolotl** [comment]
also rebased off of master to resolve the merge conflict + cleaned up commits as i'm not sure if merge+squashing is a thing in this repo

**Comment by luser** [comment]
> I could take a crack at solving it that way (figuring out a way to read the system info) in a follow-up PR, if you feel that this is best figured out sooner than later.

Given the well-sourced information about the `age` field never being zero in MSVC-produced PDBs I think your changes here are fine without any follow-up work. Feel free to merge as-is, and we won't worry about making that change unless it becomes desirable for some other reason in the future.


> also rebased off of master to resolve the merge conflict + cleaned up commits as i'm not sure if merge+squashing is a thing in this repo

We haven't enforced any particular commit hygiene in this repo thus far. As long as CI is passing and someone has approved the changes that's good enough for me.

**Comment by Gankra** [comment]
I have actually started more aggressively using "squash and merge", my policy of "eh, it's fine" has started to make going through the history a bit unwieldy (and I have been joyed to learn I can add my own notes to the commit message).

**Comment by Gankra** [comment]
Thanks for soldiering through this messy/pedantic process!

**Comment by luser** [comment]
> The more thorough approach sounds like a pretty massive undertaking. It's pretty unlikely that I'll be tackling this (heck, I'm not sure where I'd get started talking to somebody at Google about changing crashpad) but I'd hate to see this buried in this PR. Thoughts?

I'll move that novella of a comment into its own issue in this repo (I should train myself to do that from the start). I'm also not sure it's worth all the work, but given the flexibility of the minidump format and the existing Breakpad/Crashpad extensions to it, it seems silly to continue wedging bits of information like Mach-O UUIDs into data structures designed for PDB files.

The tooling for dealing with non-Windows minidumps from Breakpad has always been lacking. This change does set things up for improvement on that front—you could imagine writing a tool to fetch dSYMs for a minidump using Apple's [`DebugSymbols.framework`](https://lldb.llvm.org/use/symbols.html) now that you can correctly extract Mach-O UUIDs!

**Pull request merged**
